### PR TITLE
Transfer contract implementation

### PR DESCRIPTION
**Pull request by vlopes11**

### CODE REVIEW

File: contracts/transfer/src/ops.rs
Comment (CPerezz) [inline contracts/transfer/src/ops.rs:13]: Since these are exported, please add some docs for the module as well as the constants or file a tech-debt issue about it.

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer.rs:22]: Can't you simply type-alias? Derive all of these things for a new struct is even more code that we add to the `wasm` module.

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer.rs:32]: I'd rename to `TransferContract` to avoid any missconceptions with the methods. Maybe just `Contract` and allow the user to alias it if needed.

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer.rs:60]: This will be removed once the spurious functions are implemented properly? 
If so, file a tech-debt about removing that please.

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer.rs:45]: Why this function even exist? It does not appear in the specs and the root of the tree should only be updated by the vm once the changes to the state are applied.

File: contracts/transfer/src/wasm/bridge.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/bridge.rs:61]: File a tech-debt for that please.

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:116]: File a tech-debt for this issue or mention this in the issue created for the tech-debt generated by this PR merge.

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:191]: Ditto.

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:249]: Ditto

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:231]: Ditto

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:292]: ditto

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:314]: ditto

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:330]: ditto

File: contracts/transfer/src/wasm/spurious.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/spurious.rs:20]: Is `unwrap_or_default` the desired behavior for sure?

**Review by CPerezz — CHANGES_REQUESTED** [review]
For me it's fine to merge this but first I would require a couple of things:
- File all the issues about the tech-debts that this PR introduces.
- Remove `contracts/transfer/transfer_contract.wasm` from the files included.
- There's again no documentation at all as it happened with `transfer-circuits`. It's ok since we want the Node team to be able to integrate ASAP but we should file tech-debt issues for the docs of both crates.

Once these things are done, I'll approve the PR once a re-review is requested.

File: contracts/transfer/src/wasm/spurious.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/spurious.rs:20]: Yep. The function will return `0` for a non-existent balance

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs]: Outdated; solved with last commit

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:238]: Outdated; solved with last commit

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:296]: The pending part here is the proof verification

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs]: Outdated; solved with last commit

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:386]: Proof verification

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:49]: The `Store` update and `root` mappings update after execute are different routines.

This function exists to keep the root of the tree updated

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:64]: This is not related to spurious functions but to contract initialization. It won't be removed after the spurious functions are implemented

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [reply contracts/transfer/src/transfer.rs:49]: But aren't you assuming here that nodes will order the tx's on the same way? Idk wether that;s an assumption that can be done now whether it affects this TBH.

File: contracts/transfer/src/ops.rs
Comment (vlopes11) [reply contracts/transfer/src/ops.rs:13]: https://github.com/dusk-network/rusk/issues/191

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:26]: Fixed

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:36]: Agreed, fixed

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:49]: I don't see any potential problem out of this.

Internally, I have a set of old roots. I need to update this set whenever a successful tx is executed and mutates the notes tree

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:278]: https://github.com/dusk-network/rusk/issues/192

File: contracts/transfer/src/wasm/bridge.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/bridge.rs:61]: https://github.com/dusk-network/rusk/issues/193

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:296]: https://github.com/dusk-network/rusk/issues/194

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:348]: https://github.com/dusk-network/rusk/issues/197

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:386]: https://github.com/dusk-network/rusk/issues/194

**Review by CPerezz — APPROVED** [review]
LGTM

**Pull request merged**
